Add client-diversity page to /developer/docs (#5218)
ethereum/ethereum-org-website: src/content/developers/docs/nodes-and-clients/client-diversity/index.md

---
title: Client Diversity
description: A high level explanation of the importance of Ethereum client diversity.
lang: en
sidebar: true
sidebarDepth: 2
---

## Introduction {#Introduction}

> The behaviour of an Ethereum node is controlled by the client software it runs. There are several production-level Ethereum clients, each one developed and maintained in different languages by separate teams. The clients are built to a common spec that ensures the clients seamlessly communicate with each other and have the same functionality and provide an equivalent user experience. Multiple clients exist because the development community has decided to build multiple implementations as a way to strengthen the network. However, at the moment the distribution of clients across nodes is not even enough to realise this network fortification to its full potential. Ideally, users divide roughly equally across the various clients to bring as much client diversity as possible to the network.

Author: ok, sure- coming in next commit.

---

> title: Client Diversity

Reviewer: ```suggestion
title: Client diversity
```

## Prerequisites {#prerequisites}

If you don't already have an understanding of what nodes and clients are, check out [Nodes and clients](/developers/docs/nodes-and-clients/). The Beacon Chain is explained [here](/src/content/eth2/beacon-chain/index.md). Execution and consensus layers are defined in the [glossary](/src/content/glossary/index.md).

## Current Client Diversity? {#current-client-diversity}

![Diagram showing snapshot of client diversity](../client-diversity.jpg)
_This diagram shows snapshots of client diversity replotted using data from Ethernodes.org (execution layer) and from Michael Sproul (https://github.com/sigp/blockprint) for the consensus layer_

> The two pie charts above show snapshots of the current client diversity for the execution and consensus layers (at time of writing in January 2022). The execution layer is overwhelmingly dominated by [Geth](https://geth.ethereum.org/), with [Open Ethereum](https://openethereum.github.io/) a distant second, [Erigon](https://github.com/ledgerwatch/erigon) third and [Nethermind](https://nethermind.io/) fourth, with other clients comprising less than 1 % of the network. The most commonly used client on the consensus layer - [Prysm](https://prysmaticlabs.com/#projects) - is not as dominant as Geth but still represents over 60% of the network. [Lighthouse](https://lighthouse.sigmaprime.io/) and [Teku](https://consensys.net/knowledge-base/ethereum-2/teku/) make up ~20% and ~14% respectively, and other clients are rarely used.

Author: other option is embedding some live data from clientdiversity.org's api (for consensus layer only atm), but I'll delegate this decision up to the web team who might actually be able to implement this!

> The two pie charts above show snapshots of the current client diversity for the execution and consensus layers (at time of writing in January 2022). The execution layer is overwhelmingly dominated by [Geth](https://geth.ethereum.org/), with [Open Ethereum](https://openethereum.github.io/) a distant second, [Erigon](https://github.com/ledgerwatch/erigon) third and [Nethermind](https://nethermind.io/) fourth, with other clients comprising less than 1 % of the network. The most commonly used client on the consensus layer - [Prysm](https://prysmaticlabs.com/#projects) - is not as dominant as Geth but still represents over 60% of the network. [Lighthouse](https://lighthouse.sigmaprime.io/) and [Teku](https://consensys.net/knowledge-base/ethereum-2/teku/) make up ~20% and ~14% respectively, and other clients are rarely used.

Reviewer 2: Yeah, I guess I have some hesitation in sticking a "Current client diversity" section in the middle of this explainer. 

Client diversity as a topic is important in general... feel like we should focus on the broader topic at first, then get into this further down. Perhaps move this section just above the "Use a minority client" section below? This could help organize the page into the info that changes less or more frequently, so it's easier to maintain in the future.

> The two pie charts above show snapshots of the current client diversity for the execution and consensus layers (at time of writing in January 2022). The execution layer is overwhelmingly dominated by [Geth](https://geth.ethereum.org/), with [Open Ethereum](https://openethereum.github.io/) a distant second, [Erigon](https://github.com/ledgerwatch/erigon) third and [Nethermind](https://nethermind.io/) fourth, with other clients comprising less than 1 % of the network. The most commonly used client on the consensus layer - [Prysm](https://prysmaticlabs.com/#projects) - is not as dominant as Geth but still represents over 60% of the network. [Lighthouse](https://lighthouse.sigmaprime.io/) and [Teku](https://consensys.net/knowledge-base/ethereum-2/teku/) make up ~20% and ~14% respectively, and other clients are rarely used.

Reviewer: I agree with this. It makes the page more maintainable and makes more sense to me from an information architecture point of view.

1. **High-level understanding:** Why are there multiple clients?
2. **Potential problems:** Why is client diversity important?
3. **The state of play:** Current client diversity
4. **This solution:** Use a minority client 

In the current structure, points 2 and 3 are reversed which means people are reading about the current state of play without really understanding why this might be bad.

Although these are unlikely scenarios, the risk associated with them can be mitigated by evening out the distribution of clients across the active nodes. Ideally, no consensus client would ever have more than 33% share of the total nodes.

At the same time, there is a human cost to single client dominance because it puts excess strain and responsibility on one small team of developers. The smaller the client diversity, the greater the consequences of a mistake for the developers maintaining the dominant client. Spreading this load across multiple teams is good for the health of Ethereum's network of nodes and network of people.

## Use a minority client {use-minority-client}

> Addressing client diversity requires more than individual users to choose minority clients - it requires mining/validator pools and institutions like the major dapps and exchanges to switch clients too. However, all users can do their part in redressing the current imbalance and normalising the use of all the available Ethereum software. After the merge, all node operators will be required to run an execution client and a consensus client. Choosing combinations of the clients suggested below will help increase client diversity.

Author: @minimalsm also added reference to Shanghai here in the "bugs" section, but happy to delete if you think it is superfluous.

> ### Execution Clients

Reviewer: ```suggestion
### Execution clients {#execution-clients}
```

[Besu](https://www.hyperledger.org/use/besu)

[Nethermind](https://downloads.nethermind.io/)

[Erigon](https://github.com/ledgerwatch/erigon)

[CoreGeth](https://core-geth.org/)

> ### Consensus Clients

Reviewer: ```suggestion
### Consensus clients {#consensus-clients}
```

[Nimbus](https://nimbus.team/)

[Lighthouse](https://github.com/sigp/lighthouse)

[Teku](https://consensys.net/knowledge-base/ethereum-2/teku/)

> and soon [Lodestar](https://github.com/ChainSafe/lodestar)

Reviewer: Where are we at with this?

Author: afaik lodestar has been running very nicely for quite a while now but is not being promoted as production-grade yet until it's been battle tested a bit more. But someone more clued-in can hopefully advise.

Reviewer 2: This is my understanding as well... Haven't heard any further updates on this recently

Author: Have asked about this on Eth R&D discord and will await advice!

Reviewer 2: Heard back in Discord... gonna lean towards labeling Lodestar as under review and audit.

```suggestion
[Lodestar](https://github.com/ChainSafe/lodestar) - _Under review and audit_
```

This was also noted in #5242 ([see comment](https://github.com/ethereum/ethereum-org-website/pull/5242#discussion_r794154382))

> Technical users can help accelerate this process by writing more tutorials and documentation for minority clients and encouraging their node-operating peers to migrate away from the dominant clients. There are guides for switching to a minority consensus client available at [clientdiversity.org](https://clientdiversity.org/).

Reviewer: order here feels weird (having the list of clients in the middle of this section). just me?

---
title: Client diversity
description: A high level explanation of the importance of Ethereum client diversity.
lang: en
sidebar: true
sidebarDepth: 2
---

> The behaviour of an Ethereum node is controlled by the client software it runs. There are several production-level Ethereum clients, each one developed and maintained in different languages by separate teams. The clients are built to a common spec that ensures the clients seamlessly communicate with each other and have the same functionality and provide an equivalent user experience. However, at the moment the distribution of clients across nodes is not even enough to realise this network fortification to its full potential. Ideally, users divide roughly equally across the various clients to bring as much client diversity as possible to the network.

Reviewer 2: Per https://ethereum.org/en/contributing/style-guide/#american-english
```suggestion
The behavior of an Ethereum node is controlled by the client software it runs. There are several production-level Ethereum clients, each one developed and maintained in different languages by separate teams. The clients are built to a common spec that ensures the clients seamlessly communicate with each other and have the same functionality and provide an equivalent user experience. However, at the moment the distribution of clients across nodes is not even enough to realize this network fortification to its full potential. Ideally, users divide roughly equally across the various clients to bring as much client diversity as possible to the network.
```

## Prerequisites {#prerequisites}

> If you don't already have an understanding of what nodes and clients are, check out [Nodes and clients](/developers/docs/nodes-and-clients/). The Beacon Chain is explained [here](/src/content/eth2/beacon-chain/index.md). Execution and consensus layers are defined in the [glossary](/src/content/glossary/index.md).

Reviewer 2: These terms are now added, so I used their direct links instead. 
(Btw, links should just be the URL path after the language, so for example `https://ethereum.org/en/glossary/` can just be shortened to `/glossary/`... don't want to include the /src/content/ though)

```suggestion
If you don't already have an understanding of what nodes and clients are, check out [nodes and clients](/developers/docs/nodes-and-clients/). The Beacon Chain is explained [here](/upgrades/beacon-chain/). [Execution](/glossary/#execution-layer) and [consensus]/glossary/#consensus-layer) layers are defined in the glossary.
```

## Why are there multiple clients?

> Multiple, independently developed and maintained clients exist because client diversity makes the network more resilient to attacks and bugs. Multiple clients is a strength unique to Ethereum - other blockchains rely on the infallibility of a single client. However, it is not enough simply to have multipe, clients available, they have to be adopted by the community and the total active nodes distributed relatively evenly across them.

Reviewer 2: To emphasize the point of "why", may be worth making a general note here that if a network has one client and there is a bug, this could be catastrophic, while a bug in one client out of many would likely be insignificant.

### 2) Resilience to attacks 
Client diversity also offers resilience to attacks. For example, an attack that [tricks a particular client](https://twitter.com/vdWijden/status/1437712249926393858) onto a particular branch of the chain is unlikely to be successful because other clients are unlikely to be exploitable in the same way and the canonical chain remains uncorrupted. Low client diversity increases the risk associated with a hack on the dominant client. Client diversity has already proven to be an important defence against malicious attacks on the network, for example the Shanghai DDOS attack in 2016 was possible because attackers were able to trick the dominant client (Geth) into executing a slow disk i/o operation tens of thousands of times per block. Because alternative clients were also online which did not share the vulnerability, Ethereum was able to resist the attack and continue to operate while the vulnerability in Geth was fixed.

> ### 3) Proof-of-stake finality
> Ethereum has had 100% uptime since the network began. After the merge, the risks caused by poor client diversity become more alarming. A critical bug in a consensus client with over 33% of the Ethereum nodes could prevent the Beacon Chain from finalizing, causing Ethereum to go offline. Worse still, a critical bug in a client with a two-thirds majority could cause the chain to [incorrectly split and finalize](https://www.symphonious.net/2021/09/23/what-happens-if-beacon-chain-consensus-fails/), leading to a large set of validators getting stuck on an invalid chain. If they want to rejoin the correct chain, these validators face slashing or a slow and expensive voluntary withdrawal and reactivation. The magnitude of a slashing scales with the number of culpable nodes with a two-thirds majority slashed maximally (32 ETH).

Reviewer 2: Would argue this is one of the more threatening pieces here, and one of the most important points as to why consensus layer client diversity specifically is so important. 

To sum this up, if there is a bug in a supermajority client, then over 2/3 of the network are at risk of slashing and losing 100% of their staked ETH, which at today's levels would be a loss of about $15 billion.

Guess my point is, wondering how we could highlight this comment a little more. Anyone running a validator should be aware that it's _not_ in their best interest to be running the majority client. As in the above scenario, those running a minority client would be unaffected, and a similar bug in a minority client would result in negligible lose by comparison.

Maybe as simple as breaking this line out and adding a 🚨?

```suggestion
### 3) Proof-of-stake finality
Ethereum has had 100% uptime since the network began. After the merge, the risks caused by poor client diversity become more alarming. A critical bug in a consensus client with over 33% of the Ethereum nodes could prevent the Beacon Chain from finalizing, causing Ethereum to go offline.

<Emoji text="🚨" mr="1rem" /> Worse still, a critical bug in a client with a two-thirds majority could cause the chain to [incorrectly split and finalize](https://www.symphonious.net/2021/09/23/what-happens-if-beacon-chain-consensus-fails/), leading to a large set of validators getting stuck on an invalid chain. If they want to rejoin the correct chain, these validators face slashing or a slow and expensive voluntary withdrawal and reactivation. The magnitude of a slashing scales with the number of culpable nodes with a two-thirds majority slashed maximally (32 ETH).
```

Reviewer: We can't add an emoji inline as it breaks the subsequent markdown in this paragraph.

<img width="662" alt="Screenshot 2022-01-28 at 11 26 55" src="https://user-images.githubusercontent.com/62268199/151539485-80427682-9a07-462e-b404-35b5270c4107.png">

But agree with your point that we could bring more attention to this. One option might be to have a callout to this point in the intro paragraph? Not sure if that would be too much without the context laid out here.

Reviewer 2: Ah right... so you can, you just need to use an html anchor tag instead of markdown for the link. We do this on our history page quite a bit. 

I'm gonna add this back and keep it simple for now. Overall structure looks good as is, think we can discuss if we want to pull this into a callout in the future.

Although these are unlikely scenarios, the Ethereum eco-system can mitigate their risk by evening out the distribution of clients across the active nodes. Ideally, no consensus client would ever have more than a 33% share of the total nodes.

### 4) Shared responsibility

There is also a human cost to having majority clients. It puts excess strain and responsibility on a small development team. The lesser the client diversity, the greater the burden of responsibility for the developers maintaining the majority client. Spreading this responsibility across multiple teams is good for both the health of Ethereum's network of nodes and its network of people.

## Use a minority client {use-minority-client}

> Addressing client diversity requires more than individual users to choose minority clients - it requires mining/validator pools and institutions like the major dapps and exchanges to switch clients too. However, all users can do their part in redressing the current imbalance and normalising the use of all the available Ethereum software. After the merge, all node operators will be required to run an execution client and a consensus client. Choosing combinations of the clients suggested below will help increase client diversity.

Reviewer 2: ```suggestion
Addressing client diversity requires more than individual users to choose minority clients - it requires mining/validator pools and institutions like the major dapps and exchanges to switch clients too. However, all users can do their part in redressing the current imbalance and normalizing the use of all the available Ethereum software. After the merge, all node operators will be required to run an execution client and a consensus client. Choosing combinations of the clients suggested below will help increase client diversity.
```

And yeah, I think we should encourage a few ways for the community to get involved. 

Home stakers - Should be encourages within their abilities to chose a minority client. Ideally we have that clientdiversity.org link close by here.
Pooled stakers - Find our which client your provider is using, and if they're using the majority client, express support for switching to a minority client
Everyone else - Similarly, for the health of the network, encourage everyone to get involved in the discussion and/or education surrounding the topic. This is not straight forward and we need as much help as possible in spreading the facts about why this is so important.  Maybe some discord links, or link to the online communities page we have.

[Erigon](https://github.com/ledgerwatch/erigon)

[CoreGeth](https://core-geth.org/)

### Consensus clients {#consensus-clients}

[Nimbus](https://nimbus.team/)

[Lighthouse](https://github.com/sigp/lighthouse)

[Teku](https://consensys.net/knowledge-base/ethereum-2/teku/)

and soon [Lodestar](https://github.com/ChainSafe/lodestar)

Technical users can help accelerate this process by writing more tutorials and documentation for minority clients and encouraging their node-operating peers to migrate away from the dominant clients. Guides for switching to a minority consensus client are available on [clientdiversity.org](https://clientdiversity.org/).

## Further reading {#further-reading}

- [List of Ethereum node services](https://ethereumnodes.com/)
- [Importance of client diversity](https://our.status.im/the-importance-of-client-diversity/)
- ["Five Why's" of the client diversity problem](https://notes.ethereum.org/@afhGjrKfTKmksTOtqhB9RQ/BJGj7uh08)
- [Ethereum Diversity and How to Solve For It (Youtube)](https://www.youtube.com/watch?v=1hZgCaiqwfU)
- [clientdiversity.org](https://clientdiversity.org/)

## Related topics {#related-topics}

> - [Nodes and clients](/developers/docs/nodes-and-clients/)

Reviewer 2: New page =)
```suggestion
- [Run an Ethereum node](/run-a-node/)
- [Nodes and clients](/developers/docs/nodes-and-clients/)
```

---
title: Client diversity
description: A high level explanation of the importance of Ethereum client diversity.
lang: en
sidebar: true
sidebarDepth: 2
---

The behavior of an Ethereum node is controlled by the client software it runs. There are several production-level Ethereum clients, each one developed and maintained in different languages by separate teams. The clients are built to a common spec that ensures the clients seamlessly communicate with each other and have the same functionality and provide an equivalent user experience. However, at the moment the distribution of clients across nodes is not even enough to realize this network fortification to its full potential. Ideally, users divide roughly equally across the various clients to bring as much client diversity as possible to the network.

## Prerequisites {#prerequisites}

If you don't already have an understanding of what nodes and clients are, check out [nodes and clients](/developers/docs/nodes-and-clients/). The Beacon Chain is explained [here](/upgrades/beacon-chain/). [Execution](/glossary/#execution-layer) and [consensus]/glossary/#consensus-layer) layers are defined in the glossary.

> ## Why are there multiple clients?

Reviewer: ```suggestion
## Why are there multiple clients? {#why-multiple-clients}
```

Up to date client diversity data for the consensus layer is now available at https://clientdiversity.org/.

## Why is client diversity important? {#client-diversity-importance}

Having many independently developed and maintained clients is vital for the health of a decentralized network. Let's explore the reasons why.

> ### 1) Bugs

Reviewer: I think we should remove numbering here. To some it might imply a level of importance

```suggestion
### Bugs {#bugs}
```

A bug in an individual client is less of a risk to the network when representing a minority of Ethereum nodes. With a roughly even distribution of nodes across many clients, the likelihood of most clients suffering from a shared issue is small, and as a result, the network is more robust.

> ### 2) Resilience to attacks 

Reviewer: ```suggestion
### Resilience to attacks {#resilience} 
```

Client diversity also offers resilience to attacks. For example, an attack that [tricks a particular client](https://twitter.com/vdWijden/status/1437712249926393858) onto a particular branch of the chain is unlikely to be successful because other clients are unlikely to be exploitable in the same way and the canonical chain remains uncorrupted. Low client diversity increases the risk associated with a hack on the dominant client. Client diversity has already proven to be an important defence against malicious attacks on the network, for example the Shanghai DDOS attack in 2016 was possible because attackers were able to trick the dominant client (Geth) into executing a slow disk i/o operation tens of thousands of times per block. Because alternative clients were also online which did not share the vulnerability, Ethereum was able to resist the attack and continue to operate while the vulnerability in Geth was fixed.

> ### 3) Proof-of-stake finality

Reviewer: ```suggestion
### Proof-of-stake finality {#finality}
```

Ethereum has had 100% uptime since the network began. After the merge, the risks caused by poor client diversity become more alarming. A critical bug in a consensus client with over 33% of the Ethereum nodes could prevent the Beacon Chain from finalizing, causing Ethereum to go offline.

<Emoji text="🚨" mr="1rem" /> Worse still, a critical bug in a client with a two-thirds majority could cause the chain to [incorrectly split and finalize](https://www.symphonious.net/2021/09/23/what-happens-if-beacon-chain-consensus-fails/), leading to a large set of validators getting stuck on an invalid chain. If they want to rejoin the correct chain, these validators face slashing or a slow and expensive voluntary withdrawal and reactivation. The magnitude of a slashing scales with the number of culpable nodes with a two-thirds majority slashed maximally (32 ETH).

Although these are unlikely scenarios, the Ethereum eco-system can mitigate their risk by evening out the distribution of clients across the active nodes. Ideally, no consensus client would ever have more than a 33% share of the total nodes.

> ### 4) Shared responsibility

Reviewer: ```suggestion
### Shared responsibility {#responsibility}
```

### Execution clients {#execution-clients}

[Besu](https://www.hyperledger.org/use/besu)

[Nethermind](https://downloads.nethermind.io/)

[Erigon](https://github.com/ledgerwatch/erigon)

[CoreGeth](https://core-geth.org/)

### Consensus clients {#consensus-clients}

[Nimbus](https://nimbus.team/)

[Lighthouse](https://github.com/sigp/lighthouse)

[Teku](https://consensys.net/knowledge-base/ethereum-2/teku/)

and soon [Lodestar](https://github.com/ChainSafe/lodestar)

Technical users can help accelerate this process by writing more tutorials and documentation for minority clients and encouraging their node-operating peers to migrate away from the dominant clients. Guides for switching to a minority consensus client are available on [clientdiversity.org](https://clientdiversity.org/).

## Further reading {#further-reading}

- [List of Ethereum node services](https://ethereumnodes.com/)
- [Importance of client diversity](https://our.status.im/the-importance-of-client-diversity/)

> - ["Five Why's" of the client diversity problem](https://notes.ethereum.org/@afhGjrKfTKmksTOtqhB9RQ/BJGj7uh08)

Reviewer: I know this is the title of this post but the grammar is incorrect here.

```suggestion
- ["Five Whys" of the client diversity problem](https://notes.ethereum.org/@afhGjrKfTKmksTOtqhB9RQ/BJGj7uh08)
```

- [Ethereum Diversity and How to Solve For It (Youtube)](https://www.youtube.com/watch?v=1hZgCaiqwfU)

> - [clientdiversity.org](https://clientdiversity.org/)

Reviewer: Should we make this the top furhter reading? Seems like the logical next step to redpill people on client diversity.

### Proof-of-stake finality {#finality}
Ethereum has had 100% uptime since the network began. After the merge, the risks caused by poor client diversity become more alarming. A critical bug in a consensus client with over 33% of the Ethereum nodes could prevent the Beacon Chain from finalizing, causing Ethereum to go offline.

> Worse still, a critical bug in a client with a two-thirds majority could cause the chain to [incorrectly split and finalize](https://www.symphonious.net/2021/09/23/what-happens-if-beacon-chain-consensus-fails/), leading to a large set of validators getting stuck on an invalid chain. If they want to rejoin the correct chain, these validators face slashing or a slow and expensive voluntary withdrawal and reactivation. The magnitude of a slashing scales with the number of culpable nodes with a two-thirds majority slashed maximally (32 ETH).

Reviewer 2: ```suggestion
<Emoji text="🚨" mr="1rem" /> Worse still, a critical bug in a client with a two-thirds majority could cause the chain to <a href="https://www.symphonious.net/2021/09/23/what-happens-if-beacon-chain-consensus-fails/" target="_blank">incorrectly split and finalize</a>, leading to a large set of validators getting stuck on an invalid chain. If they want to rejoin the correct chain, these validators face slashing or a slow and expensive voluntary withdrawal and reactivation. The magnitude of a slashing scales with the number of culpable nodes with a two-thirds majority slashed maximally (32 ETH).
```
